File: Replace/MIN/block_version/min.py
from collections import defaultdict, OrderedDict, deque

# 原论文中的版本


class MINCache(object):
    def __init__(self, capacity):
        self.capacity = capacity
        self.evicted = deque()  # 浏览一次访问流后得到输入集最大化命中率的替换顺序
        # {index: {id: True, id: True}} 记录缓存中各个序号有多少块
        self.request_index = defaultdict(OrderedDict)
        self.block = {}  # {id: index}
        self.complete = 1
        self.cur = 0

        self.map = {}  # {id: True} 真实缓存空间，在第二次访问流时才使用

    def train(self, id):
        # 该块未被访问过或已经被替换
        if id not in self.block or self.block[id] < self.complete:
            self.cur += 1
            self.block[id] = self.cur
            self.request_index[self.cur][id] = True
            return

        if self.block[id] == self.cur:  # 连续访问
            return

        if self.block[id] < self.cur and self.block[id] >= self.complete:  # 仍然在缓存中，且重用距离大于1
            self.request_index[self.block[id]].pop(id)
            if not self.request_index[self.block[id]]:
                del self.request_index[self.block[id]]
            self.block[id] = self.cur
            self.request_index[self.cur][id] = True

            temp = self.cur
            counter = 0

            while True:
                if self.request_index[temp]:
                    counter += len(self.request_index[temp])
                if counter == self.capacity or temp == self.complete:
                    self.complete = temp  # 索引小于complete的块消失
                    no_value = []
                    for idx, d in self.request_index.items():
                        if idx >= self.complete:
                            for t in no_value:
                                del self.request_index[t]
                            return
                        for evic, b in d.items():
                            self.evicted.append(evic)
                        no_value.append(idx)

                if counter < self.capacity:
                    temp -= 1
                    counter -= 1
                    continue

    def build_tail(self):
        for idx, d in self.request_index.items():
            for evic, b in d.items():
                self.evicted.append(evic)

    def visit(self, id):
        if self.capacity == 0:
            return

        if id in self.map:
            return

        if len(self.map) < self.capacity:
            self.map[id] = True
            return

        key = self.evicted.popleft()
        del self.map[key]
        self.map[id] = True


print("Enter MINCache Capacity (Blocks) : ", end='')
capa = int(input())
print("Enter <Your Test File>: ", end='')
file = input()
cache = MINCache(capa)

with open(file, "r", encoding="utf-8") as f:
    all_visit = 0
    hit = 0
    line = f.readline()
    while line:
        temp = line.split()
        id = int(temp[1])
        cache.train(id)
        line = f.readline()
    cache.build_tail()
f.close()

with open(file, "r", encoding="utf-8") as f:
    all_visit = 0
    hit = 0
    line = f.readline()
    while line:
        temp = line.split()
        id = int(temp[1])
        if id in cache.map:
            hit += 1
        cache.visit(id)
        all_visit += 1
        line = f.readline()
    print("-------MIN替换算法-------")
    print("缓存容量 (Blocks) : " + str(capa))
    print("总请求数: " + str(all_visit) + ", 命中次数: " + str(hit))
    print("-------MIN替换算法-------")
f.close()

File: Replace/MIN/size_version/min.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <vector>
#include <string.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <set>
#include <list>
#include <queue>
#include <climits>
using namespace std;

class MINCache //opt算法占用空间开销大，因此 key 和 value 只设置为 int 类型了
{
private:
    long long int capacity, cache_size = 0;
    long long int total_num = 0, hit_num = 0;
    unordered_map<int, int> cache;              //id->size
    unordered_map<int, queue<int>> &id_indexes; //id: 出现在哪些行

public:
    MINCache(long long int _capacity, unordered_map<int, queue<int>> &_id_indexes) : capacity(_capacity), id_indexes(_id_indexes)
    {
        cache.clear();
    }

    void visit(int key, int value)
    {
        total_num++;
        id_indexes[key].pop();

        if (cache.count(key))
        {
            hit_num++;
        }
        else
        {
            if (value > capacity)
                return;
            cache[key] = value;
            cache_size += value;
            while (cache_size > capacity)
            {
                int evicted = searchMaxReuseId();
                cache_size -= cache[evicted];
                cache.erase(evicted);
            }
        }
        cout << total_num << endl;
    }

    long long int getTotal() const
    {
        return total_num;
    }

    long long int getHit() const
    {
        return hit_num;
    }

private:
    int searchMaxReuseId() const
    {
        int max_line = -1, evicted = -1;
        for (const pair<int, int> &block : cache) //选择前向重用距离最大的id
        {
            int id = block.first;
            const queue<int> &q = id_indexes.at(id);
            if (q.front() > max_line)
            {
                max_line = q.front();
                evicted = id;
            }
        }
        return evicted;
    }
};

void makeReusePos(string file_in, string file_out)
{
    fstream fin(file_in);
    string line, temp;
    stringstream splitline;
    unordered_map<int, vector<int>> reusePos;
    reusePos.clear();
    int idx = 1;
    while (getline(fin, line)) //遍历每个文件出现的行索引，加入到id_indexes[key]中
    {
        vector<string> data;
        splitline.clear();
        splitline.str(line);
        while (getline(splitline, temp, ' '))
        {
            data.push_back(temp);
        }
        int key = stol(data[1]);
        if (!reusePos.count(key))
        {
            reusePos[key] = vector<int>();
        }
        reusePos[key].push_back(idx);
        idx++;
    }
    string tmp_file = file_out + ".tmp";
    ofstream fout(tmp_file);
    for (const pair<int, vector<int>> &pr : reusePos) //将每个id出现的行位置输出到文件中
    {
        fout << pr.first << " : ";
        for (int pos : pr.second)
        {
            fout << pos << " ";
        }
        fout << endl;
    }
    fin.close();
    fout.close();
    if (rename(tmp_file.c_str(), file_out.c_str()))
    {
        cout << "Exception in renaming file from " << tmp_file << " to " << file_out << " code: " << strerror(errno) << endl;
        exit(-1);
    }
}

int main()
{
    string line, temp;
    stringstream splitline;

    long long int capa;
    cout << "Enter MINCache Capacity (Bytes) : ";
    cin >> capa;
    string test;
    cout << "Enter <Your Test File>: ";
    cin >> test;

    cout << "Making ReusePosition File..." << endl;
    ifstream f(test + ".pos");
    if (!f.good()) //检查文件是否存在
    {
        makeReusePos(test, test + ".pos");
    }
    cout << "ReusePosition File Complete" << endl;

    fstream f_index(test + ".pos");
    unordered_map<int, queue<int>> id_indexes;
    id_indexes.clear();
    while (getline(f_index, line)) //遍历每个文件出现的行索引，加入到id_indexes[key]中
    {
        vector<string> data;
        splitline.clear();
        splitline.str(line);
        while (getline(splitline, temp, ' '))
        {
            data.push_back(temp);
        }
        int key = stol(data[0]);
        id_indexes[key] = queue<int>();
        for (int i = 2; i < data.size(); i++)
        {
            id_indexes[key].push(stol(data[i]));
        }
        id_indexes[key].push(INT32_MAX); //最后加一个无穷大
    }
    cout << id_indexes[32].size() << endl;

    fstream fin_test(test);
    MINCache cache(capa, id_indexes);

    while (getline(fin_test, line))
    {
        splitline.clear();
        splitline.str(line);
        vector<string> data;
        while (getline(splitline, temp, ' '))
        {
            data.push_back(temp);
        }

        //检查是否有某个块比缓存容量还大
        int key = stol(data[1]), value = stol(data[0]);

        cache.visit(key, value);
    }
    cout << "-------MIN替换算法-------" << endl;
    cout << "缓存容量 (Bytes) ：" << capa << endl;
    cout << "总请求数: " << cache.getTotal() << ", 命中次数: " << cache.getHit() << endl;
    cout << "-------MIN替换算法-------" << endl;

    f_index.close();
    fin_test.close();
}
